### ISSUE TITLE
Support for Erlang 20.0

### ISSUE THREAD

**Issue by GalaxyGorilla** [esl/MongooseIM#1352]
Hey, do you plan to release MIM 2.1.0 with support for Erlang 20? Having a glance on the dependencies I can see that this is not just about MIM :-/.

**Comment by fenek** [comment]
We'd definitely like to ensure 20.0 by the time 2.1.0 is release but unfortunately I can't promise it now. We need to assess how much work is required to fix dependencies and (if necessary) MIM code itself.

**Comment by deadjdona** [comment]
Well, it's possible to build with OTP20, but lot's of warnings...

**/home/ubuntu/MongooseIM# ./rebar3 compile**
```
compiling with FIPS mode enabledcompiling with FIPS mode enabled===> Verifying dependencies...
===> Skipping goldrush (from {pkg,<<"goldrush">>,<<"0.1.7">>,
                                     <<"349A351D17C71C2FDAA18A6C2697562ABE136FEC945F147B381F0CF313160228">>}) as an app of the same name has already been fetched
===> Generating ASN.1 files.
===> Generating ASN.1 files.
===> Compiling ejabberd
apps/ejabberd/src/mod_aws_sns.erl:295: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1

_build/default/lib/ejabberd/src/rdbms/mongoose_rdbms.erl:399: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

_build/default/lib/ejabberd/src/http_upload/mod_http_upload.erl:173: Warning: variable 'GetUrl' is unused

apps/ejabberd/src/mongoose_cassandra.erl:228: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1
apps/ejabberd/src/mongoose_cassandra.erl:262: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1

apps/ejabberd/src/mod_muc_room.erl:180: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3
apps/ejabberd/src/mod_muc_room.erl:186: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3
apps/ejabberd/src/mod_muc_room.erl:196: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:206: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:217: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:228: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:774: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/mod_muc_room.erl:1881: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/bin_to_hex.erl:none: Warning: this system is not configured for native-code compilation.

apps/ejabberd/src/mod_bosh_socket.erl:99: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3
apps/ejabberd/src/mod_bosh_socket.erl:131: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/mod_bosh_socket.erl:138: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/mod_bosh_socket.erl:145: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:149: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:154: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:159: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:164: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:771: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/ejabberd_sm.erl:148: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)
apps/ejabberd/src/ejabberd_sm.erl:159: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

apps/ejabberd/src/ejabberd_local.erl:156: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

apps/ejabberd/src/extauth.erl:137: Warning: random:seed/1: the 'random' module is deprecated; use the 'rand' module instead
apps/ejabberd/src/extauth.erl:138: Warning: random:uniform/1: the 'random' module is deprecated; use the 'rand' module instead

apps/ejabberd/src/ejabberd_s2s.erl:614: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2

apps/ejabberd/src/ejabberd_check.erl:34: Warning: export_all flag enabled - all functions will be exported

apps/ejabberd/src/mod_carboncopy.erl:59: Warning: type matchspec_atom() is unused

apps/ejabberd/src/randoms.erl:34: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1

apps/ejabberd/src/ejabberd_router.erl:450: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)
apps/ejabberd/src/ejabberd_router.erl:463: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

apps/ejabberd/src/mod_muc.erl:797: Warning: gen_fsm:sync_send_all_state_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/mod_muc.erl:813: Warning: gen_fsm:sync_send_all_state_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/mod_muc.erl:1046: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/ejabberd_receiver.erl:259: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/ejabberd_receiver.erl:310: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/ejabberd_receiver.erl:329: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/ejabberd_receiver.erl:347: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/ejabberd_s2s_out.erl:1108: Warning: random:seed/1: the 'random' module is deprecated; use the 'rand' module instead
apps/ejabberd/src/ejabberd_s2s_out.erl:1113: Warning: random:uniform/0: the 'random' module is deprecated; use the 'rand' module instead

apps/ejabberd/src/eldap.erl:188: Warning: gen_fsm:start_link/4 is deprecated and will be removed in a future release; use gen_statem:start/4
apps/ejabberd/src/eldap.erl:195: Warning: gen_fsm:start_link/4 is deprecated and will be removed in a future release; use gen_statem:start/4
apps/ejabberd/src/eldap.erl:204: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/eldap.erl:212: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/eldap.erl:232: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:258: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:275: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:320: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:329: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:344: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:401: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:778: Warning: gen_fsm:reply/2 is deprecated and will be removed in a future release; use gen_statem:reply/2
apps/ejabberd/src/eldap.erl:807: Warning: gen_fsm:reply/2 is deprecated and will be removed in a future release; use gen_statem:reply/2

apps/ejabberd/src/acl.erl:127: Warning: variable 'Default' is unused
apps/ejabberd/src/acl.erl:129: Warning: variable 'Default' is unused

apps/ejabberd/src/ejabberd_listener.erl:422: Warning: variable 'Module' is unused

apps/ejabberd/src/cyrsasl_oauth.erl:19: Warning: variable 'Host' is unused

apps/ejabberd/src/ejabberd_s2s_in.erl:131: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3
```
**/home/ubuntu/MongooseIM# make rel**
```
. ./configure.out && ./rebar3 as prod release
compiling with FIPS mode enabledcompiling with FIPS mode enabledcompiling with FIPS mode enabled===> Verifying dependencies...
===> Skipping goldrush (from {pkg,<<"goldrush">>,<<"0.1.7">>,
                                     <<"349A351D17C71C2FDAA18A6C2697562ABE136FEC945F147B381F0CF313160228">>}) as an app of the same name has already been fetched
===> Generating ASN.1 files.
===> Generating ASN.1 files.
===> Compiling ejabberd
apps/ejabberd/src/mod_aws_sns.erl:295: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1

_build/prod/lib/ejabberd/src/rdbms/mongoose_rdbms.erl:399: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

_build/prod/lib/ejabberd/src/http_upload/mod_http_upload.erl:173: Warning: variable 'GetUrl' is unused

apps/ejabberd/src/mongoose_cassandra.erl:228: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1
apps/ejabberd/src/mongoose_cassandra.erl:262: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1

apps/ejabberd/src/mod_muc_room.erl:180: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3
apps/ejabberd/src/mod_muc_room.erl:186: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3
apps/ejabberd/src/mod_muc_room.erl:196: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:206: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:217: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:228: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_muc_room.erl:774: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/mod_muc_room.erl:1881: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/bin_to_hex.erl:none: Warning: this system is not configured for native-code compilation.

apps/ejabberd/src/mod_bosh_socket.erl:99: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3
apps/ejabberd/src/mod_bosh_socket.erl:131: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/mod_bosh_socket.erl:138: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/mod_bosh_socket.erl:145: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:149: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:154: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:159: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:164: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/mod_bosh_socket.erl:771: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/ejabberd_sm.erl:148: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)
apps/ejabberd/src/ejabberd_sm.erl:159: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

apps/ejabberd/src/ejabberd_local.erl:156: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

apps/ejabberd/src/extauth.erl:137: Warning: random:seed/1: the 'random' module is deprecated; use the 'rand' module instead
apps/ejabberd/src/extauth.erl:138: Warning: random:uniform/1: the 'random' module is deprecated; use the 'rand' module instead

apps/ejabberd/src/ejabberd_s2s.erl:614: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2

apps/ejabberd/src/ejabberd_check.erl:34: Warning: export_all flag enabled - all functions will be exported

apps/ejabberd/src/mod_carboncopy.erl:59: Warning: type matchspec_atom() is unused

apps/ejabberd/src/randoms.erl:34: Warning: crypto:rand_uniform/2 is deprecated and will be removed in a future release; use rand:uniform/1

apps/ejabberd/src/ejabberd_router.erl:450: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)
apps/ejabberd/src/ejabberd_router.erl:463: Warning: erlang:get_stacktrace/0 used following an old-style 'catch' may stop working in a future release. (Use it inside 'try'.)

apps/ejabberd/src/mod_muc.erl:797: Warning: gen_fsm:sync_send_all_state_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/mod_muc.erl:813: Warning: gen_fsm:sync_send_all_state_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/mod_muc.erl:1046: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/ejabberd_receiver.erl:259: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/ejabberd_receiver.erl:310: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/ejabberd_receiver.erl:329: Warning: gen_fsm:send_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/ejabberd_receiver.erl:347: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1

apps/ejabberd/src/ejabberd_s2s_out.erl:1108: Warning: random:seed/1: the 'random' module is deprecated; use the 'rand' module instead
apps/ejabberd/src/ejabberd_s2s_out.erl:1113: Warning: random:uniform/0: the 'random' module is deprecated; use the 'rand' module instead

apps/ejabberd/src/eldap.erl:188: Warning: gen_fsm:start_link/4 is deprecated and will be removed in a future release; use gen_statem:start/4
apps/ejabberd/src/eldap.erl:195: Warning: gen_fsm:start_link/4 is deprecated and will be removed in a future release; use gen_statem:start/4
apps/ejabberd/src/eldap.erl:204: Warning: gen_fsm:sync_send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:call/2
apps/ejabberd/src/eldap.erl:212: Warning: gen_fsm:send_all_state_event/2 is deprecated and will be removed in a future release; use gen_statem:cast/1
apps/ejabberd/src/eldap.erl:232: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:258: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:275: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:320: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:329: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:344: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:401: Warning: gen_fsm:sync_send_event/3 is deprecated and will be removed in a future release; use gen_statem:call/3
apps/ejabberd/src/eldap.erl:778: Warning: gen_fsm:reply/2 is deprecated and will be removed in a future release; use gen_statem:reply/2
apps/ejabberd/src/eldap.erl:807: Warning: gen_fsm:reply/2 is deprecated and will be removed in a future release; use gen_statem:reply/2

apps/ejabberd/src/acl.erl:127: Warning: variable 'Default' is unused
apps/ejabberd/src/acl.erl:129: Warning: variable 'Default' is unused

apps/ejabberd/src/ejabberd_listener.erl:422: Warning: variable 'Module' is unused

apps/ejabberd/src/cyrsasl_oauth.erl:19: Warning: variable 'Host' is unused

apps/ejabberd/src/ejabberd_s2s_in.erl:131: Warning: gen_fsm:start_link/3 is deprecated and will be removed in a future release; use gen_statem:start/3

===> Starting relx build process ...
===> Resolving OTP Applications from directories:
          /home/ubuntu/MongooseIM/_build/prod/lib
          /home/ubuntu/MongooseIM/apps
          /usr/lib/erlang/lib
          /home/ubuntu/MongooseIM/_build/prod/rel
===> Resolved mongooseim-2.1.0beta2
===> Including Erts from /usr/lib/erlang
===> release successfully created!

```

**Comment by fenek** [comment]
OTP 20 is supported in 2.1.0 release.
